src/noit_filters.c

```diff
@@ -506,18 +494,24 @@ noit_apply_filterset(const char *filterset,
       }
       /* If we need some of these and we have an auto setting that isn't fulfilled for each of them, we can add and succeed */
 #define CHECK_ADD(rname) (!need_##rname || (r->rname##_auto_hash_max > 0 && r->rname##_ht && mtev_hash_size(r->rname##_ht) < r->rname##_auto_hash_max))
-      if(CHECK_ADD(target) && CHECK_ADD(module) && CHECK_ADD(name) && CHECK_ADD(metric)) {
-#define UPDATE_FILTER_RULE(rnum, rname, value) do { \
-  mtev_hash_replace(r->rname##_ht, strdup(value), strlen(value), NULL, free, NULL); \
-  if(noit_filter_update_conf_rule(fs->name, rnum, #rname, value) < 0) { \
-    mtevL(noit_error, "Error updating configuration for new filter auto_add on %s=%s\n", #rname, value); \
+#define UPDATE_FILTER_RULE(rname, value) do { \
+  if(r->flush_interval.tv_sec || r->flush_interval.tv_usec) { \
+    struct timeval reset; \
+    add_timeval(r->last_flush, r->flush_interval, &reset); \
+    if(compare_timeval(now, reset) >= 0) { \
+      mtev_hash_delete_all(r->rname##_ht, free, NULL); \
+      flushed = mtev_true; \
+    } \
   } \
+  mtev_hash_replace(r->rname##_ht, strdup(value), strlen(value), NULL, free, NULL); \
 } while(0)
-        if(need_target) UPDATE_FILTER_RULE(idx, target, check->target);
-        if(need_module) UPDATE_FILTER_RULE(idx, module, check->module);
-        if(need_name) UPDATE_FILTER_RULE(idx, name, check->name);
-        if(need_metric) UPDATE_FILTER_RULE(idx, metric, metric->metric_name);
-        noit_filterset_log_auto_add(fs->name, check, metric, r->type == NOIT_FILTER_ACCEPT);
+      if(CHECK_ADD(target) && CHECK_ADD(module) && CHECK_ADD(name) && CHECK_ADD(metric)) {
+        mtev_boolean flushed = mtev_false;
+        if(need_target) UPDATE_FILTER_RULE(target, check->target);
+        if(need_module) UPDATE_FILTER_RULE(module, check->module);
+        if(need_name) UPDATE_FILTER_RULE(name, check->name);
+        if(need_metric) UPDATE_FILTER_RULE(metric, metric->metric_name);
+        if(flushed) memcpy(&r->last_flush, &now, sizeof(now));
```

Comment:
How does `r->last_flush` get initialized or is `add_timeval()` okay with that?

Reply:
It is initialized by the initial calloc of the rule (to zero)